### ISSUE TITLE
Apply Aggregate is broken

### ISSUE THREAD

**Issue by ocharles** [circuithub/rel8#36]
Thanks to help from @tomjaguarpaw, I think I can finally conclude that `Apply Aggregate` won't work:

```haskell
aggregate $ values [lit True, lit False] >>= \x -> pure ((x &&.) <$> Rel8.or x)
```

produces

```sql
SELECT
("values0_1") AND ("result0_2") as "anon"
FROM (SELECT
      *
      FROM (SELECT
            BOOL_OR("inner0_2") as "result0_2"
            FROM (SELECT
                  "values0_1" as "inner0_2",
                  *
                  FROM (SELECT
                        *
                        FROM (SELECT "column1" as "values0_1"
                              FROM
                              (VALUES
                               (CAST(TRUE AS bool)),
                               (CAST(FALSE AS bool))) as "V") as "T1") as "T1") as "T1"
            GROUP BY COALESCE(0)) as "T1") as "T1"
```

which errors with

```
ERROR:  column "values0_1" does not exist
LINE 2: ("values0_1") AND ("result0_2") as "anon"
```

There is a clear scoping violation in that query. When we had `Applicative Aggregate` with the traversal to group anything unaggregated this was OK, but then we found out you can't group arbitrary expressions (the type _must_ have a notion equality).

I think we've reached the end of this exploration. Oh well!

**Comment by ocharles** [comment]
Actually, I'm back to thinking we could get all the way back to `Applicative`:

I wrote on Matrix:

> The broken values` [lit True, lit False] >>= \x -> aggregate (values [x, x] >>= \y -> pure ((y &&.) <$> Rel8.or y))` query should mean the following SQL:
>
> ```sql
> select b.x OR c.x from (values (true), (false)) a(x), lateral (values (a.x), (a.x)) b(x), lateral (select bool_or(q.x) from (values (b.x)) q(x)) c(x);.
> ```
>
> Basically anything that you have in an `Aggregate` that's not aggregated is an implicit group by, but that group by isn't `GROUP BY`, but is instead pulled out of the query entirely and the aggregation becomes a lateral aggregation, once for each row
>
> Just need to figure out a way to do this rewriting...
>
> It means that the result of the above would be four rows:
>
> ```
> ?column? 
>----------
> t
> t
> f
> f
> (4 rows)
> ```
> 
> It's a bit tricky, but if you are aggregating a `Query (Aggregate a)`, you need to look inside that `Aggregate a` and find all `BaseTableAttrs` that aren't aggregated, and then trace them back to the query they came from.

**Comment by ocharles** [comment]
Fixed in #47
